File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if (bloomDay.length < m * k) return -1;
        
        // return makeBouquets(3, bloomDay, m, k);
        int min = Integer.MAX_VALUE;
        int max = 0;
        int ans = -1;
        for (int i=0; i < bloomDay.length; i++) {
            min = Math.min(min, bloomDay[i]);
            max = Math.max(max, bloomDay[i]);
        }

        while (min <= max) {
            int mid = (min + max) / 2;
            int days = makeBouquets(mid, bloomDay, m, k);
            if (days == -1) {
                min = mid + 1;
            }
            else {
                ans = mid;
                max = mid - 1;
            }
        }

        return ans;

    }

    static int makeBouquets(int day, int[] bloomDay, int m, int k) {
        int makeM = 0;
        int makeK = 0;

        for (int i=0; i < bloomDay.length; i++) {
            if (makeK == k){
                makeM ++;
                makeK = 0;
            }
            
            if (bloomDay[i] > day) {
                makeK = 0;
            } else {
                makeK ++;
            }
        }
        if (makeK == k){
                makeM ++;
                makeK = 0;
        }

        // System.out.println(makeM);
        if (makeM >= m) return day;
        else return -1;
    }
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
// #define long long int long long int
class Solution {
public:
    bool makeBouquets(long long int day, long long int m, long long int k, vector<int> &bloomDay) {
        int n = bloomDay.size();
        int i = 0;
        while ( i < n ) {
            long long int adj = 0;
            while (i < n && bloomDay[i] <= day) {
                // cout << "hi  " << i << endl;
                adj++;
                i++;
                if (adj == k) {
                    i--;
                    break;
                }
            }
            // cout << "adj " << adj << endl;
            if (adj == k) {
                m -= 1;
                if (m == 0) break;
            }
            i++;
        }
        cout << m << endl;
        return m == 0;        
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        long long int n = bloomDay.size();
        long long int ans = INT_MAX;
        if (n <  (long long int)m * (long long int)k) return -1;

        long long int left = 1, right = *max_element(bloomDay.begin(), bloomDay.end());

        while ( left <= right ) {
            long long int mid = (left + right) /2;
            // cout << mid << " " << makeBouquets(mid, m, k, bloomDay) << endl;
            if ( makeBouquets(mid, (long long int)m, (long long int)k, bloomDay) ) {
                ans = min(ans, mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        if (ans == INT_MAX) return -1;
        return ans;


    }
};